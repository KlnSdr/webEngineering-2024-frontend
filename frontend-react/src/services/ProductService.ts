import {NeededProduct, Product} from "../types/Products";

class ProductsService {
  private static backendURL: string =
    process.env.REACT_APP_BACKEND_URL || "http://localhost:13000";

  public static getAll(): Promise<Product[]> {
    return new Promise((resolve, reject) => {
        fetch(`${this.backendURL}/products`).then((response: Response) => {
            if (!response.ok) {
                throw new Error("Failed to load products.");
            }
            return response.json();
        }).then((data: Product[]) => {
            console.log(data);
            resolve(data);
        }).catch((reason: any) => {
            reject(reason);
        });
    });
  }
  
  //get Id from product
    public static getProductIdByName(productName: string): Promise<NeededProduct | null> {
        return this.getAll().then(products => {
            const product = products.find(prod => prod.name.toLowerCase() === productName.toLowerCase());
            return product ? { id: product.id, productName: product.name, amount: 0 } : null;
        });
    }

}

export { ProductsService };
