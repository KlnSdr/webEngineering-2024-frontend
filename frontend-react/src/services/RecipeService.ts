import { CreateRecipe } from "../types/Recipes";
import { Recipe } from "../types/Recipes";
import {NeededProduct} from "../types/Products";

class RecipeService {
  private static backendURL: string =
    process.env.REACT_APP_BACKEND_URL || "http://localhost:13000";

  public static save(recipe: CreateRecipe): Promise<void> {
    return new Promise((resolve, reject) => {
      fetch(`${this.backendURL}/recipes`, {
        method: "POST",
        credentials: "include",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(recipe),
      })
        .then((response: Response) => {
          if (!response.ok) {
            throw new Error("Could not save recipe.");
          }
          return response.json();
        })
        .then((data: any) => {
          console.log(data);
          resolve();
        })
        .catch((reason: any) => {
          reject(reason);
        });
    });
  }
  public static getAll1(): Promise<CreateRecipe[]> {
    return new Promise((resolve, reject) => {
      resolve([
        {
          title: "Käsesoße 1",
          image: null,
          description: "Description 1",
          products: [],
        },
        {
          title: "Käsesoße 2",
          image: null,
          description: "Description 2",
          products: [],
        },
        {
          title: "Käsesosse 3",
          image: null,
          description: "Description 3",
          products: [],
        },
        {
          title: "Käsesosse 4",
          image: null,
          description: "Description 3",
          products: [],
        },
      ]);
    });
  }

  public static getAll():Promise<Recipe[]> {
    return new Promise((resolve, reject) => {
      fetch(`${this.backendURL}/recipes`).then((response: Response) => {
          if (!response.ok) {
            throw new Error("Failed to load recipes.");
          }
          return response.json();
        }).then((data: Recipe[]) => {
          resolve(data);
        })
        .catch((reason: any) => {
          reject(reason);
        });
    });
  }

    public static getAll(): Promise<CreateRecipe[]> {
        return new Promise((resolve, reject) => {
            resolve([
                {
                    title: "Käsesoße 1",
                    image: null,
                    description: "A delicious cheese sauce.",
                    products: [
                        { id: 7, productName: "Cheese", amount: 1 },
                        { id: 4, productName: "Milk", amount: 1 },
                        { id: 6, productName: "Butter", amount: 1 },
                    ],
                },
                {
                    title: "Käsesoße 2",
                    image: null,
                    description: "Cheese sauce with salt.",
                    products: [
                        { id: 7, productName: "Cheese", amount: 2 },
                        { id: 10, productName: "Salt", amount: 1 },
                        { id: 4, productName: "Milk", amount: 2 },
                    ],
                },
                {
                    title: "Käsesoße 3",
                    image: null,
                    description: "A spicy cheese sauce.",
                    products: [
                        { id: 7, productName: "Cheese", amount: 2 },
                        { id: 11, productName: "Pepper", amount: 1 },
                        { id: 2, productName: "Milk", amount: 1 },
                    ],
                },
                {
                    title: "Käsesoße 4",
                    image: null,
                    description: "A mild cheese sauce.",
                    products: [
                        { id: 7, productName: "Cheese", amount: 3 },
                        { id: 2, productName: "Milk", amount: 2 }
                    ],
                },
            ]);
        });
    }


    public static searchRecipesByProducts(neededProducts: NeededProduct[]): Promise<CreateRecipe[]> {
        return new Promise((resolve, reject) => {
            this.getAll()
                .then((recipes) => {

                    const staticRecipes: CreateRecipe[] = [
                        {
                            title: "Käsesoße 1",
                            image: null,
                            description: "A delicious cheese sauce.",
                            products: [
                                { id: 7, productName: "Cheese", amount: 1 },
                                { id: 4, productName: "Milk", amount: 1 },
                                { id: 6, productName: "Butter", amount: 1 },
                            ],
                        },
                        {
                            title: "Käsesoße 2",
                            image: null,
                            description: "Cheese sauce with salt.",
                            products: [
                                { id: 7, productName: "Cheese", amount: 2 },
                                { id: 10, productName: "Salt", amount: 1 },
                                { id: 4, productName: "Milk", amount: 2 },
                            ],
                        },
                        {
                            title: "Käsesoße 3",
                            image: null,
                            description: "A spicy cheese sauce.",
                            products: [
                                { id: 7, productName: "Cheese", amount: 2 },
                                { id: 11, productName: "Pepper", amount: 1 },
                                { id: 2, productName: "Milk", amount: 1 },
                            ],
                        },
                        {
                            title: "Käsesoße 4",
                            image: null,
                            description: "A mild cheese sauce.",
                            products: [
                                { id: 7, productName: "Cheese", amount: 3 },
                                { id: 2, productName: "Milk", amount: 2 }
                            ],
                        },
                    ];

                    resolve(staticRecipes);
                })
                .catch((reason: any) => {
                    reject(reason);
                });
        });
    }

}

export { RecipeService };
