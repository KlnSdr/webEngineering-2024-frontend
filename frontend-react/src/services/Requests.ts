// just one more abstraction bro :D

/**
 * Sends a request to the specified path with the given options.
 * @param {string} path - The URL path to send the request to.
 * @param {Object} [options={}] - The options to use for the request.
 * @returns {Promise<Response>} A promise that resolves to the response of the request.
 */
function request(path: string, options: any = {}): Promise<Response> {
    return fetch(path, options);
}

/**
 * Sends an authorized request to the specified path with the given options.
 * Adds an Authorization header with a Bearer token if available.
 * @param {string} path - The URL path to send the request to.
 * @param {Object} [options={}] - The options to use for the request.
 * @returns {Promise<Response>} A promise that resolves to the response of the request.
 */
function authorizedRequest(path: string, options: any = {}): Promise<Response> {
    const token = localStorage.getItem("token");
    if (token) {
        options.headers = {
            ...options.headers,
            Authorization: `Bearer ${token}`,
        };
    }
    return request(path, options);
}

export { request, authorizedRequest };