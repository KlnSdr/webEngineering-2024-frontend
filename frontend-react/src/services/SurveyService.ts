import {CreateSurvey, Survey} from "../types/Surveys";;

class SurveyService {
    private static backendURL: string =
        process.env.REACT_APP_BACKEND_URL || "http://localhost:13000";

    public static getSurveysByUserId(userId: string): Promise<any[]> {
        return new Promise((resolve, reject) => {
            fetch(`${this.backendURL}/surveys/user/${userId}`).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Failed to load surveys.");
                }
                return response.json();
            }).then((data: Survey[]) => {
                resolve(data);
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }

    public static getSurveyById(surveyId: string | undefined): Promise<Survey> {
        return new Promise((resolve, reject) => {
            fetch(`${this.backendURL}/surveys/${surveyId}`, {credentials: "include"}).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Failed to load survey.");
                }
                return response.json();
            }).then((data: Survey) => {
                console.log(data);
                resolve(data);
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }
    public static voteSurvey(surveyId: number, recipeId: number): Promise<Survey>{
        return new Promise((resolve, reject) =>{
            fetch(`${this.backendURL}/surveys/${surveyId}/vote/${recipeId}`, {
                method: "PUT",
                credentials: "include",
                headers: {
                    "Content-Type": "application/json",
                },
            }).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Failed to vote.");
                }
                return response.json();
            }).then((data: Survey) => {
                resolve(data);
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }

    public static createSurvey(survey: CreateSurvey): Promise<void> {
        const surveyToSend = {
            title: survey.title,
            options: survey.options.map((recipe: any) => `/recipes/${recipe.id}`),
        };

        return new Promise((resolve, reject) => {
            fetch(`${this.backendURL}/surveys`, {
                method: "POST",
                credentials: "include",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(surveyToSend),
            }).then((response: Response) => {
                if (!response.ok) {
                    throw new Error("Could not save survey.");
                }
                return response.json();
            }).then((data: any) => {
                console.log(data);
                resolve();
            }).catch((reason: any) => {
                reject(reason);
            });
        });
    }
}

export default SurveyService;