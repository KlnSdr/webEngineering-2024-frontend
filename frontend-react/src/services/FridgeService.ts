import { NeededProduct } from "../types/Products";
import {ProductsService} from "./ProductService";


class FridgeService {
    private static backendURL: string =
        process.env.REACT_APP_BACKEND_URL || "http://localhost:13000";

    public static getFridgeContent(userId: number): Promise<NeededProduct[]> {
        return new Promise((resolve, reject) => {
            fetch(`${this.backendURL}/fridge/${userId}`, {
                method: "GET",
                credentials: "include",
            })
                .then((response) => {
                    if (!response.ok) {
                        throw new Error("Failed to load fridge content.");
                    }
                    return response.json();
                })
                .then(async (data: { userUri: string; products: Record<string, number> }) => {
                    const products: NeededProduct[] = [];

                    // Fetch product IDs for each product name
                    for (const [productName, amount] of Object.entries(data.products)) {
                        const productId = await ProductsService.getProductIdByName(productName);
                        if (productId !== null) {
                            products.push({
                                id: productId.id,
                                productName: productId.productName,
                                amount,
                            });
                        } else {
                            console.warn(`Product ID not found for ${productName}`);
                        }
                    }

                    resolve(products);
                })
                .catch((reason: any) => {
                    reject(reason);
                });
        });
    }

    public static updateFridgeContent(userId: number, products: NeededProduct[]): Promise<void> {
        return new Promise((resolve, reject) => {
            // Convert the products array to a map
            const productsMap: Record<string, number> = {};
            products.forEach((product) => {
                productsMap[product.productName] = product.amount;
            });

            fetch(`${this.backendURL}/fridge/${userId}`, {
                method: "PUT",
                credentials: "include",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({ userUri: `/users/${userId}`, products: productsMap }),
            })
                .then((response) => {
                    if (!response.ok) {
                        throw new Error("Failed to update fridge content.");
                    }
                    resolve();
                })
                .catch((reason: any) => {
                    reject(reason);
                });
        });
    }

}

export { FridgeService };
