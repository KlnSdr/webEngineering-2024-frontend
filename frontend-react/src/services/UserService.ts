import {CurrentUser, User} from "../types/Users";
import {jwtDecode} from "jwt-decode";
import {request} from "./Requests";

/**
 * Service class for handling user-related operations.
 */
class UserService {
  /**
   * The backend URL for the user service.
   * @type {string}
   * @private
   */
  private static backendURL: string =
    process.env.REACT_APP_BACKEND_URL || "http://localhost:13000";

  /**
   * Cached user information.
   * @type {CurrentUser | null}
   * @private
   */
  private static userInfo: CurrentUser | null = null;

  /**
   * Logs out the current user by removing the token and redirecting to the logout URL.
   */
  static doLogout() {
    localStorage.removeItem("token");
    window.location.assign(`${UserService.backendURL}/users/logout`);
  }

  /**
   * Initiates the login process by redirecting to the OAuth2 authorization URL.
   */
  static doLogin() {
    window.location.assign(
      `${UserService.backendURL}/oauth2/authorization/gitlab`
    );
  }

  /**
   * Retrieves the current user's information.
   * @returns {Promise<any>} A promise that resolves to the current user's information.
   */
  static getUserInfo(): Promise<any> {
    return new Promise<any>((resolve, reject) => {
      if (UserService.userInfo !== null) {
        resolve(UserService.userInfo);
        return;
      }

      if (localStorage.getItem("token")) {
          this.loadUserInfo(localStorage.getItem("token")!);
      }

      resolve(UserService.userInfo);
    });
  }

  /**
   * Retrieves a user by their URI.
   * @param {string} uri - The URI of the user.
   * @returns {Promise<User>} A promise that resolves to the user.
   */
  static getUser(uri: string): Promise<User> {
      return new Promise((resolve, reject) => {
          request(`${this.backendURL}${uri}`, {})
          .then(response => {
              if (!response.ok) {
                  throw new Error("could not get user with uri " + uri);
              }
              return response.json();
          })
          .then(resolve)
          .catch(_ => {reject()});
      });
  }

  /**
   * Checks if the user is logged in.
   * @returns {Promise<boolean>} A promise that resolves to a boolean indicating if the user is logged in.
   */
  static isLoggedIn(): Promise<boolean> {
    return new Promise<boolean>((resolve) => {
      UserService.init().then((_) => resolve(UserService.userInfo !== null));
    });
  }

  /**
   * Loads the user information from the given token.
   * @param {string} token - The JWT token.
   * @private
   */
  static loadUserInfo(token: string) {
    UserService.userInfo = jwtDecode<CurrentUser>(token);
  }

  /**
   * Initializes the user service by loading the user information.
   * @returns {Promise<void>} A promise that resolves when the initialization is complete.
   * @private
   */
  private static init(): Promise<void> {
    return new Promise((resolve) => {
      UserService.getUserInfo()
        .then((_) => resolve())
        .catch((_) => resolve());
    });
  }
}

export { UserService };