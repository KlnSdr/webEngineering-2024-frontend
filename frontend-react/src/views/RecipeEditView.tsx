import React, { useEffect, useState } from 'react';
import {Recipe} from "../types/Recipes";
import {useParams} from "react-router-dom";
import CreateRecipeView from "./CreateRecipeView";
import {Product} from "../types/Products";
import {ProductsService} from "../services/ProductService";
import {authorizedRequest} from "../services/Requests";

function EditRecipeView() {
    const {id} = useParams<{id:string}>();
    const [recipe, setRecipe] = useState<Recipe | null>(null);
    const backendURL: string = process.env.REACT_APP_BACKEND_URL || "http://localhost:13000";

    useEffect(() => {
        ProductsService.getAll().then((productDetails: Product[]) => {
            (async () => {
                try {
                    const response = await authorizedRequest(`${backendURL}/recipes/${id}`);
                    if (!response.ok) {
                        throw new Error("Failed to load recipe.");
                    }
                    const data: any = await response.json();
                    const recipe: Recipe = {
                        id: data.id,
                        title: data.title,
                        description: data.description,
                        imgUri: data.imgUri,
                        isPrivate: data.isPrivate,
                        creationDate: new Date(data.creationDate),
                        ownerUri: data.ownerUri,
                        likedByUserUris: data.likedByUserUris,
                        products: Object.keys(data.productQuantities).map((key: string) => {
                            const productId: number = parseInt(key.replace("/products/", ""));
                            const productDetail: Product | undefined = productDetails.find((val: Product) => val.id === productId);
                            return {
                                id: productId,
                                name: productDetail?.name || "---",
                                amount: data.productQuantities[key],
                                unit: productDetail?.unit || "---"
                            };
                        })
                    };
                    setRecipe(recipe);
                } catch (error) {
                    console.error(error);
                }
            })();
        });
    }, [id, backendURL]);


    if (recipe === null) {
        return <div>Rezept mit Id {id} konnte nicht gefunden werden.</div>
    } else {
        return (<div>
            <CreateRecipeView recipe={recipe} />
        </div>);
    }


}

export default EditRecipeView;