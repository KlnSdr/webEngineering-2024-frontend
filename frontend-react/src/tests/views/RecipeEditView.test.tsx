import React from "react";
import { render, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import {UserService} from "../../services/UserService";
import RecipeEditView from "../../views/RecipeEditView";

jest.mock("../../services/UserService");

const mockIsLoggedIn = UserService.isLoggedIn as jest.Mock;

describe("RecipeEditView Component", () => {

    test('matches snapshot for RecipeEditView component if user is logged in', () => {
        mockIsLoggedIn.mockResolvedValue(true);
        const { asFragment } = render(
            <MemoryRouter initialEntries={["/recipes/edit/:id"]}>
                <RecipeEditView />
            </MemoryRouter>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('matches snapshot for RecipeEditView component if user is not logged in', () => {
        mockIsLoggedIn.mockResolvedValue(false);
        const { asFragment } = render(
            <MemoryRouter initialEntries={["/recipes/edit/:id"]}>
                <RecipeEditView />
            </MemoryRouter>
        );

        expect(asFragment()).toMatchSnapshot();
    });

});
