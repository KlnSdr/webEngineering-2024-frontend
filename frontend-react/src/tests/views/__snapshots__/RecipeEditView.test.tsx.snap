// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeEditView Component matches snapshot for RecipeEditView component if user is logged in 1`] = `
<DocumentFragment>
  <div>
    Rezept mit Id  konnte nicht gefunden werden.
  </div>
</DocumentFragment>
`;

exports[`RecipeEditView Component matches snapshot for RecipeEditView component if user is not logged in 1`] = `
<DocumentFragment>
  <div>
    Rezept mit Id  konnte nicht gefunden werden.
  </div>
</DocumentFragment>
`;
